File: api/src/models/minio.model.ts
import { Client } from 'minio';

export class MinIOModel {
  private client: Client;
  private bucket: string;

  constructor() {
    this.client = new Client({
      endPoint: process.env.MINIO_ENDPOINT!,
      port: Number(process.env.MINIO_PORT),
      useSSL: false,
      accessKey: process.env.MINIO_ACCESS_KEY!,
      secretKey: process.env.MINIO_SECRET_KEY!,
    });

    this.bucket = process.env.MINIO_BUCKET_NAME!;
  }

  async listFiles(prefix: string): Promise<string[]> {
    const objects: string[] = [];

    return new Promise((resolve, reject) => {
      const stream = this.client.listObjectsV2(this.bucket, prefix, true);

      stream.on('data', (obj) => {
        if (obj.name) objects.push(obj.name);
      });

      stream.on('end', () => resolve(objects));
      stream.on('error', reject);
    });
  }

  async getFile(objectName: string): Promise<Buffer> {
    const stream = await this.client.getObject(this.bucket, objectName);

    return new Promise((resolve, reject) => {
      const chunks: Buffer[] = [];

      stream.on('data', (chunk) => chunks.push(chunk));
      stream.on('end', () => resolve(Buffer.concat(chunks)));
      stream.on('error', reject);
    });
  }
}

File: api/backend/models/minio_model.py
from __future__ import annotations

from io import BytesIO

from minio import Minio


class MinIOModel:
    def __init__(self, *, endpoint: str, port: int, access_key: str, secret_key: str, bucket: str, use_ssl: bool = False) -> None:
        self.client = Minio(
            endpoint=f"{endpoint}:{port}",
            access_key=access_key,
            secret_key=secret_key,
            secure=use_ssl,
        )
        self.bucket = bucket

    def list_files(self, prefix: str) -> list[str]:
        objects = self.client.list_objects(self.bucket, prefix=prefix, recursive=True)
        return [obj.object_name for obj in objects if obj.object_name]

    def get_file(self, object_name: str) -> bytes:
        response = self.client.get_object(self.bucket, object_name)
        try:
            stream = BytesIO(response.read())
            return stream.getvalue()
        finally:
            response.close()
            response.release_conn()
